### PR TITLE
Basic Signature Verification

### PR DESCRIPTION
**Pull request by rgnote**
1. Perform signature verification
2. Support `signingAuthority` trust store as spec'd in https://github.com/notaryproject/notaryproject/pull/165/files

Signed-off-by: rgnote <5878554+rgnote@users.noreply.github.com>

### CODE REVIEW

File: verification/policy.go
Comment (gokarnm) [inline verification/policy.go:126]: Define as an enum, it's being used in a couple of places.

File: verification/verifier_helpers.go
Comment (gokarnm) [inline verification/verifier_helpers.go:23]: Sweet!! Nice to see this integrated finally @priteshbandi @rgnote !

File: verification/verifier_helpers.go
Comment (gokarnm) [inline verification/verifier_helpers.go:161]: We should return an error in this case as there is no x509 subject trusted identity to consider.

**Review by gokarnm — CHANGES_REQUESTED** [review]
Nice work keeping the code modular and readable !! 🚀

File: verification/verifier.go
Comment (rgnote) [reply verification/verifier.go:49]: Callers do have to loop through the array to figure out whether verification was successful or not. They just need to check whether the error is null or not. They need to look at the array only if they want to make use of the individual results.

File: verification/verifier.go
Comment (rgnote) [reply verification/verifier.go:74]: Will use `unable to retrieve digital signature(s) associated..`

File: verification/helpers.go
Comment (rgnote) [reply verification/helpers.go:61]: Did you mean tag support?

File: verification/verifier.go
Comment (gokarnm) [reply verification/verifier.go:49]: The CLI will always return signed annotations to caller for successful verification, so it will need to loop through this.

**Review by gokarnm — APPROVED** [review]

**Comment by gokarnm** [comment]
Pending second review and [related spec](https://github.com/notaryproject/notaryproject/pull/165) approval.

File: verification/verifier.go
Comment (SteveLasker) [inline verification/verifier.go:140]: Why do we need `notary.default.x509`, as opposed to `notary.x509`
Shouldn't default be a configuration that might change in the future?

**Review by SteveLasker — COMMENTED** [review]
This is great to start testing. I do think we should land [Verification Plugin and Signing Scheme spec #165](https://github.com/notaryproject/notaryproject/pull/165) before merging this.

File: verification/verifier.go
Comment (rgnote) [reply verification/verifier.go]: @SteveLasker I agree that this PR should not implement any points that are proposed in https://github.com/notaryproject/notaryproject/pull/165 unless it is merged. I removed the comment.

File: verification/verifier_helpers.go
Comment (shizhMSFT) [reply verification/verifier_helpers.go:62]: Please follow the golang conversion [Indent Error Flow](https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow), and apply to other code blocks.

File: verification/verifier_helpers.go
Comment (shizhMSFT) [inline verification/verifier_helpers.go:67]: This is a big TODO 💀

File: internal/mock/mocks.go
Comment (rgnote) [reply internal/mock/mocks.go:170]: I wanted this mock to be available for all of notation-go code as there are few packages that would benefit, like `signature`, `verification` and `plugin`. Do you still see any issues since this is an internal package?

File: verification/helpers.go
Comment (rgnote) [reply verification/helpers.go:58]: [oras.land/oras-go/registry.ParseReference()](https://pkg.go.dev/oras.land/oras-go/v2@v2.0.0-rc.1.0.20220727034506-eb13fdfeefa6/registry#ParseReference) supports tag references as well, which notation doesn't support yet. We can clean up this code as part of https://github.com/notaryproject/notation-go/issues/81

File: verification/policy.go
Comment (rgnote) [reply verification/policy.go:238]: Good idea. Will update it.

File: verification/verifier.go
Comment (rgnote) [reply verification/verifier.go:57]: We are skipping the error check here because this is handled at L31

File: verification/verifier.go
Comment (rgnote) [reply verification/verifier.go:65]: That's a miss. Will address it.

File: verification/verifier_helpers.go
Comment (rgnote) [reply verification/verifier_helpers.go:67]: Dir Structure was not implemented when I raised this PR. I will integrate with https://github.com/notaryproject/notation-go/tree/main/dir and remove these TODOs shortly

File: verification/verifier_helpers.go
Comment (rgnote) [reply verification/verifier_helpers.go:62]: Good idea. Will address.

File: internal/mock/mocks.go
Comment (shizhMSFT) [reply internal/mock/mocks.go:170]: I mean we could save this JSON file in the `testdata` folder like `internal/mock/testdata/envolope.json`.

File: verification/verifier.go
Comment (shizhMSFT) [reply verification/verifier.go:57]: We either need to check the error or write a comment.

File: verification/verifier.go
Comment (JeyJeyGao) [reply verification/verifier.go:27]: prerequisite #90

File: verification/verifier.go
Comment (rgnote) [reply verification/verifier.go]: Added a comment

File: verification/verifier_helpers.go
Comment (shizhMSFT) [reply verification/verifier_helpers.go:165]: It seems `github.com/notaryproject/notation-core-go/signer` will ensure that but just comment here to confirm.

**Review by shizhMSFT — APPROVED** [review]
LGTM

File: verification/verifier_helpers.go
Comment (rgnote) [reply verification/verifier_helpers.go:165]: Correct, if there is no signing cert, we wouldn't come this far to verify trusted identities.

File: verification/verifier_helpers.go
Comment (rgnote) [reply verification/verifier_helpers.go:101]: Will address this and the other nits in my next PR

**Review by gokarnm — APPROVED** [review]
LGTM for recent changes

**Pull request merged**
